fixing issue 39191

narendraingale2 opened the pull request:
### What does this PR do?

This PR is fixing issue 39191. 

### What issues does this PR fix or reference?
issue 39191

### Previous Behavior
refer this issue  https://github.com/saltstack/salt/issues/39191


### Tests written?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

bbinet committed:
Restore support to generate thin tarballs with local packages

(when absonly=False)

See: https://github.com/saltstack/salt/pull/35997#discussion_r100556792

Commit:
Added code for issue 39191

Commit:
Added code for issue 39191

Commit:
Added some comments to help to understand new code.

Commit:
Added some comments to help to understand new code.

Commit:
Insert space after comma

Required for lint check to pass

Commit:
Insert space after comma

Required for lint check to pass

cachedout commented:
@ticosax Could you please take a look here?

ticosax commented on salt/states/dockerng.py:
```diff
@@ -202,6 +202,15 @@ def _image_get(path, default=None):
+            # Comparing list vice-versa
+            for key in actual_env:
+                # Need to skip PATH and no_proxy variables.
+                if key not in ['PATH', 'no_proxy']:
```
please explain this exclusion of `PATH` and `no_proxy`.

narendraingale2 replied:
These 2 variables are set by docker engine by default. 'docker inspect' shows these variables in ENV section as below. So these variables are included in list of actual_env variables and if we don't skip them container will get restarted every time even if there is no change in user defined environment variables.

"Env": [
                "VAR3=value33",
                "no_proxy=*.local, 169.254/16",
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],

ticosax replied:
Please add in the comment that those environment variables are controlled by docker then.

ticosax commented on salt/states/dockerng.py:
```diff
@@ -202,6 +202,15 @@ def _image_get(path, default=None):
+                    if key not in data:
+                        env_ptr = env_diff.setdefault(item, {})
+                        env_ptr.setdefault('old', {})[key] = actual_val
+                        env_ptr.setdefault('new', {})[key] = None
```
I'd rather skip it the `new` entry since the variable is not defined. Having the previous entry just in `old` should be fine.
your call though.

narendraingale2 replied:
Agree I will remove new entry.

ticosax reviewed (commented):
looks good.
But I'd like to be sure why `PATH` and `no_proxy` are not compared.

ticosax commented:
I forgot, a test covering this fix would be awesome.

Commit:
Added some changes.

Commit:
Added some changes.

narendraingale2 commented:
@ticosax , @cachedout  This is my first contribution and sorry to say but I don't know how to write test cases. Though I would like to write. If you could direct me to some tutorial/video would be great. Thanks in advance.

ticosax commented on salt/states/dockerng.py:
```diff
@@ -205,12 +205,12 @@ def _image_get(path, default=None):
             # Comparing list vice-versa
             for key in actual_env:
                 # Need to skip PATH and no_proxy variables.
+                actual_val = actual_env.get(key)
```
you can do
```python
for key, actual_val in actual_env.items():
```
instead

narendraingale2 replied:
done.

ticosax commented:
@narendraingale2 everything is explained here.
https://docs.saltstack.com/en/latest/topics/development/tests/index.html

You will see dockerng coverage is pretty decent. You will be able to start easily from an existing test.

Commit:
Added few more changes.

Commit:
Added few more changes.

whiteinge committed:
Add full_return argument to LocalClient.cmd to expose exit code

whiteinge committed:
Add full_return support to RunnerClient.cmd_sync to expose exit code

whiteinge committed:
Add LocalClient and RunnerClient full_return tests

Commit:
Allow states to declare _connection_ keys as documented

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

narendraingale2 commented:
Can somebody help me to run test case . When I try to run I am getting below error. I am trying to run integration tests. 

```
sudo ../salt/tests/runtests.py --name=integration.modules.pillar
 -> integration.modules.pillar.PillarModuleTest.test_pillar_items  .............................................................................................................
       Traceback (most recent call last):
         File "/Users/ningale/salt/tests/integration/modules/pillar.py", line 121, in test_pillar_items
           get_items = self.run_function('pillar.items')
         File "/Users/ningale/salt/tests/integration/__init__.py", line 1745, in run_function
           minion_tgt, function, arg, timeout=timeout, kwarg=kwargs
         File "/Users/ningale/salt/salt/client/__init__.py", line 701, in cmd
           **kwargs)
         File "/Users/ningale/salt/salt/client/__init__.py", line 340, in run_job
           'The salt master could not be contacted. Is master running?'
       SaltClientError: The salt master could not be contacted. Is master running?
```

Commit:
Remove trailing whitespace

Commit:
Remove trailing whitespace

Commit:
Cleanup lint errors

cachedout commented:
@narendraingale2 Are you going to be able to come back and add the test requested by @ticosax ?

narendraingale2 commented:
Sure @cachedout  I was on vacation. Will submit soon.

Regards,
Narendra

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' of https://github.com/narendraingale2/salt into develop

Commit:
Added test case for docker.

Commit:
Added test case for docker.

gtmanfred committed:
add get_or_set_hash to sdb

cachedout reviewed (changes requested):
In order for our new test runner to pick up these new tests, we need test/integration/states/docker.py renamed to test_docker.py, please.

gtmanfred committed:
consolidate test of valid sdb uri

gtmanfred committed:
fix pylint

Commit:
Correct lack of kwargs

twangboy committed:
Fix incorrect test name

twangboy committed:
Fix test_worktree_add_rm test for Windows

twangboy committed:
Fix test_add_user_to_group for Windows

tkwilliams committed:
INFRA-4579 - fix some deref errors I introduced awhile back, a small docstring fix, and increase default timeouts for cache subnet group and replication group creations

s0undt3ch committed:
Normalize the mode

Commit:
renamed:    tests/integration/states/docker.py -> tests/integration/states/test_docker.py

Commit:
renamed:    tests/integration/states/docker.py -> tests/integration/states/test_docker.py

Commit:
Support the creation of .zip thin file

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge pull request #39795 from twangboy/fix_whitelist

Fix incorrect test name

Commit:
Merge pull request #39800 from twangboy/win_fix_test_useradd

Fix test_add_user_to_group for Windows

Commit:
Merge pull request #39803 from bodhi-space/infra4579

INFRA-4579 - small fixes for boto3_elasticache state module

kadefor committed:
Fix salt-ssh cp.get_url wapper bug

In fileclient.get_url function, the 2nd arg is "dest", not "saltenv".

carlpett committed:
Vault: Always lowercase policies

Commit:
Minor typo fix

Commit:
Changed "Template is an empty file" to debug.

s0undt3ch committed:
Account for test module names prefixed with `tests.`

s0undt3ch committed:
Error out when the path passed is not valid tests module

s0undt3ch committed:
Merge pull request #39816 from s0undt3ch/develop

Minor fixes to running tests

Commit:
Merge pull request #39810 from marnovdm/feature-fix-interafce-typo

Minor typo fix

s0undt3ch committed:
Gate 3rd party import

Commit:
Merge pull request #39815 from guettli/patch-2

Changed "Template is an empty file" to debug.

s0undt3ch committed:
Separate the error

s0undt3ch committed:
Minor fixes

s0undt3ch committed:
Add test case which will catch SOME problems with __virtual__ calls.

And also 3rd-party import errors.

Commit:
Merge pull request #39808 from carlpett/vault-lowercase-policies

Vault: Always lowercase policies

Commit:
Merge pull request #39809 from kadefor/patch-1

Fix salt-ssh cp.get_url wapper bug

Commit:
Merge pull request #39790 from gtmanfred/develop

add get_or_set_hash to sdb

Commit:
Merge pull request #39494 from mchugh19/conn_updates

Allow states to declare _connection_ keys as documented

cachedout commented:
@narendraingale2 We need a few more changes. First, there are a number of lint errors which need to be resolved:

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/9206/violations/file/tests/integration/states/test_docker.py/

Second, a test failed: integration.states.test_docker.TestModuleDockereng.test_docker_env

At a glance, this may be because some platforms do not have the proper prequsities installed. The test should be gated so that tests are skipped if the prereqs aren't found.

Commit:
Merge pull request #39415 from bbinet/gen_thin-absonly

Restore support to generate thin tarballs with local packages

blarghmatey committed:
Added fix for empty overrides value in apply config methods

In `apply_minion_config` and `apply_master_config` there is a call to
`overrides.get` while the `overrides` value still has the potential of
being `None`. Updated the call to be `(overrides or {}).get` to
prevent unwarranted exceptions.

whiteinge committed:
Add dummy kwarg to WheelClient().cmd_sync() to match mixin

At some point these two kwargs should be wired up to match RunnerClient.

twangboy committed:
Fix type in second worktree_path

s0undt3ch committed:
Upgrade the six library

1.10.0
------

- Issue benjaminp/six#122: Improve the performance of `six.int2byte` on Python 3.

- Pull request benjaminp/six#55 and issue benjaminp/six#99: Don't add the `winreg` module to `six.moves`
  on non-Windows platforms.

- Pull request benjaminp/six#60 and issue benjaminp/six#108: Add `six.moves.getcwd` and
  `six.moves.getcwdu`.

- Pull request benjaminp/six#64: Add `create_unbound_method` to create unbound methods.

twangboy committed:
Add encode for Py3 compatability

s0undt3ch committed:
Don't expect any ordering from sets!!!!!

s0undt3ch committed:
Lint fixes

s0undt3ch committed:
This needs to be done as a PyLint plugin not as this kind of test.

It's the only way we can check all code paths of the __virtual__
function.
Regarding the 3rd-party modules gating, we also need to create a pylint
plugin which known what are the hard deps of Salt and will complain on
all other 3rd-party imports.

s0undt3ch committed:
Merge pull request #39822 from s0undt3ch/features/py3

A tiny bit more of py3

terminalmage committed:
Fix regression in docker module when client is spawned

PR #39597 broke this module by introducing an UnboundLocalError. This
fixes that error.

Commit:
Added SkipIf for docker-py

Commit:
Added SkipIf for docker-py

Commit:
Removed new lines

Commit:
Removed new lines

Commit:
Removed white spaces

Commit:
Removed white spaces

Commit:
adjusted indentation

Commit:
adjusted indentation

moio committed:
minionswarm.py: add flag to delay starts

Commit:
Removed black listes package use.

Commit:
Removed black listes package use.

Commit:
Merge pull request #39829 from moio/minionswarm-delay-starts

minionswarm.py: add flag to delay starts

Commit:
Merge pull request #39825 from terminalmage/fix-docker-timeout

Fix regression in docker module when client is spawned

Commit:
Merge pull request #39821 from mitodl/develop

Added fix for empty overrides value in apply config methods

Commit:
Merge pull request #39797 from twangboy/win_fix_test_git

Fix test_worktree_add_rm test for Windows

Commit:
Merge pull request #39492 from whiteinge/full_return-arg

Add full_return arg to LocalClient and RunnerClient

s0undt3ch committed:
Merge pull request #39806 from cachedout/thin_zip

Support the creation of .zip thin file

Commit:
Merge branch 'develop' of https://github.com/narendraingale2/salt into develop

narendraingale2 commented:
@cachedout I am unable to run integration test cases.
`Traceback (most recent call last):
  File "../salt/tests/runtests.py", line 59, in <module>
    from tests.integration import TestDaemon  # pylint: disable=W0403
  File "/Users/ningale/salt/tests/integration/__init__.py", line 49, in <module>
    from tests.support.processes import *  # pylint: disable=wildcard-import
  File "/Users/ningale/salt/tests/support/processes.py", line 18, in <module>
    from pytestsalt.utils import SaltRunEventListener as PytestSaltRunEventListener
ImportError: No module named pytestsalt.utils
(salt-dev) ningale-L111:salt ningale$ `

Could you please help me to resolve this issue ?

Commit:
Revert "Removed black listes package use."

This reverts commit fa3052ec836356507df60c9d7fc1bf2cf1f750e7.

Commit:
revert changes

Commit:
Merge branch 'develop' of https://github.com/narendraingale2/salt into develop

narendraingale2 commented:
Sorry but I dont know how files to be changed are 41. So I closed this PR and opened new. 

https://github.com/saltstack/salt/pull/39838

sorry for inconvenience caused.
